### PR TITLE
feat: create mobile hamburger menu

### LINKED ISSUE #57: Navbar - Add links and styling to hamburger menu
# Requirements

- Add links that normally show in landscape mode to the hamburger menu. 
- Add a nice css styling using [tailwind css](http://tailwindcss.com/) best practices

----

This has a dependency on PR #56 

---- 
Design

- [Wireframe ](https://www.figma.com/file/QXQxlw4jNFIhMxkKU7bbwq/Find-a-doc%2C-Japan?type=design&node-id=1276%3A1561&mode=design&t=lXY8G3vAnxFyHqpq-1 )

### PR DESCRIPTION
**Pull request by ermish**
Resolves #57 

## 🔧 What changed
- ended up deleting the `menuStore.ts` because it was only used locally in the hamburger menu. 
- improved our media query handling for switching to mobile mode. Now using orientation vs pixel widths.  
- moved hamburger menu to the right side after convo with design team
- refactored hamburger menu
  - renamed to `HamburgerMenu`  🍔 
  - moved mobile top navbar components into `TopNav` that were in the hamburger menu. 
  - Renamed links in the menu  (About Findadoc -> About Us, Submit -> Add a Doctor)
  - added title to hamburger menu
  - created footer with terms, conditions, and dev links
  - changed (Powered by Netlify -> simpler netlify icon (should match their [open source policy](https://www.netlify.com//legal/open-source-policy/#main) 💯 
  - created close button (x) with closing functionality ❌ 
  - Added ability to close menu when you click outside. Had to build a vue custom directive  ✨ 
  - Added language selector  🇯🇵 
  - Added copyright and NPO references ©️ 
  - Added a sweet transition (thanks @Tch4lla !)

- Improved our tests
  - Added cypress recording key to the yarn script. This will let us view cypress runs in the cloud (helpful to check CI failures)
  -  make a nice `yarn test` that runs all our tests
  - removed SEMICOLONS;;;;;;;
  - fixed failing tests from hamburger changes 

## 🧪 Testing instructions
- test mobile and desktop views. 

## 📸 Screenshots

-   ### Before
portrait mode
image

landscape mode (notice it's not responsive to the monitor size)
image

-   ### After
- mobile - smallest screen
image

landscape mode - smallest screen width
image

landscape mode - ultrawide
image

### CODE REVIEW

**Comment by Tch4lla** [comment]
Are you still planning on adding a transition so that it eases in from the left?

**Comment by ermish** [comment]
> Are you still planning on adding a transition so that it eases in from the left? 

I am, but I'm also looking at swipe capability from a library 🚀

**Review by theyokohamalife — COMMENTED** [review]
This looks really great so far! I like what you did with the footer links.

You'll notice that CI is failing 2 checks. It may have something to do with the breakpoints. 
Screenshot 2024-06-03 at 7 45 59 AM

We'll want to make sure that the hamburger menu is showing for any screen size that is too small to display all of the menu items without overlapping.
Screenshot 2024-06-03 at 7 44 15 AM

**Comment by theyokohamalife** [comment]
Nit: Could you rename the PR title to something a bit more conventional? I think it would be good to avoid exclamation marks, as those are usually reserved for breaking changes in most projects.

File: components/HamburgerMenu.vue [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                        </div>
+                    </NuxtLink>
+                    <NuxtLink :to="'/about'">
+                        <div @click="toggleMenu()">
+                            {{ $t('hamburgerMenu.about') }}
+                        </div>
+                    </NuxtLink>
+                    <NuxtLink :to="'https://forms.gle/4E763qfaq46kEsn99'" target="_blank">
+                        <div @click="toggleMenu()">
+                            {{ $t('hamburgerMenu.contact') }}
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:39]: This string needs to be added to `ja.json`. 
Screenshot 2024-06-03 at 8 10 44 AM

File: components/HamburgerMenu.vue [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                </div>
+                <div id="row-2" class="flex gap-6">
+                    <NuxtLink :to="'https://github.com/ourjapanlife'" target="_blank">
+                        <SVGGithubIcon id="github-icon" alt="hamburger menu" class="h-10 w-8" />
+                    </NuxtLink>
+                    <!-- Netlify Icons are available here: https://www.netlify.com/press/#badges -->
+                    <NuxtLink :to="'https://www.netlify.com'" target="_blank">
+                        <img src="https://www.netlify.com/v3/img/components/netlify-light.svg" alt="Deploys by Netlify"
+                            class=" h-10 w-20" />
+                    </NuxtLink>
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:71]: I think it would be good to include a test that the Netlify element is always present in the hamburger menu.

File: components/HamburgerMenu.vue [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            if (!isClickInsideElement) {
+                closeMenu()
+            }
+        };
+        document.addEventListener("click", el.clickOutsideEvent)
+    },
+    unmounted: (el: any) => {
+        document.removeEventListener("click", el.clickOutsideEvent);
+    }
+}
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:107]: Nice! ✨ I like that you thought of including the click outside!

File: components/HamburgerMenu.vue [hunk 4]
```diff
@@ -0,0 +1,109 @@
+<template>
+    <div class="cursor-pointer w-full flex" v-close-on-outside-click>
+        <SVGHamburgerMenuIcon id="hamburger-menu-icon" @click="toggleMenu()" alt="hamburger menu" class="h-8 w-8" />
+        <div id="hamburger-menu" v-show="isMenuOpen"
+            class="z-20 absolute top-0 left-0 h-full w-2/3 pt-6 pb-8 rounded bg-primary-bg border-2 flex flex-col justify-between">
+            <div id="upper-section">
+                <div id="header" class="flex justify-between px-5">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:7]: Nit: I think we should make the div ids more descriptive. What do you think about something like `hamburger-menu-header` instead?

File: components/HamburgerMenu.vue [hunk 5]
```diff
@@ -0,0 +1,109 @@
+<template>
+    <div class="cursor-pointer w-full flex" v-close-on-outside-click>
+        <SVGHamburgerMenuIcon id="hamburger-menu-icon" @click="toggleMenu()" alt="hamburger menu" class="h-8 w-8" />
+        <div id="hamburger-menu" v-show="isMenuOpen"
+            class="z-20 absolute top-0 left-0 h-full w-2/3 pt-6 pb-8 rounded bg-primary-bg border-2 flex flex-col justify-between">
+            <div id="upper-section">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:6]: ♻️ 
```suggestion
            <div id="hamburger-upper-section">
```

File: components/HamburgerMenu.vue [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                <div id="header" class="flex justify-between px-5">
+                    <div class="flex">
+                        <div class="title-text flex flex-col flex-shrink-0" data-testid="logo">
+                            <div class="text-lg text-primary-text group-hover:text-primary-hover font-semibold">
+                                Find a Doc, Japan
+                            </div>
+                        </div>
+                    </div>
+                    <button id="close-button" @click="closeMenu()" class="px-2 py-.5 mr-2">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:15]: ♻️

File: components/HamburgerMenu.vue [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                        <svg class="stroke-primary" width="20" heigh="20" viewBox="4 0 15 25">
+                            <path stroke-width="3" fill="none" d="M6.25,6.25,17.75,17.75" />
+                            <path stroke-width="3" fill="none" d="M6.25,17.75,17.75,6.25" />
+                        </svg>
+                    </button>
+                </div>
+                <div id="language-section" class="flex mt-2 px-5">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:22]: ♻️

File: components/HamburgerMenu.vue [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                    <span class="text-primary-text self-center mr-2">{{ $t('hamburgerMenu.languageDropdownTitle') }}</span>
+                    <LocaleSelector class="py-1.5 text-primary-text bg-primary-bg" />
+                </div>
+                <div id="menu-items" class="mt-10 px-5 flex flex-col gap-6">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:26]: ♻️

File: components/HamburgerMenu.vue [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    <NuxtLink :to="'/submit'">
+                        <div @click="toggleMenu()">
+                            {{ $t('hamburgerMenu.submit') }}
+                        </div>
+                    </NuxtLink>
+                </div>
+            </div>
+
+            <div id="footer-section" class="flex flex-col gap-5 px-5">
+                <div id="row-1" class="flex gap-4">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:51]: ♻️

File: components/HamburgerMenu.vue [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                            {{ $t('footer.terms') }}
+                        </span>
+                    </NuxtLink>
+                    <NuxtLink :to="'/privacypolicy'">
+                        <span @click="toggleMenu()">
+                            {{ $t('footer.privacy') }}
+                        </span>
+                    </NuxtLink>
+                </div>
+                <div id="row-2" class="flex gap-6">
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:63]: ♻️

**Comment by ermish** [comment]
> This looks really great so far! I like what you did with the footer links.
> 
> You'll notice that CI is failing 2 checks. It may have something to do with the breakpoints. Screenshot 2024-06-03 at 7 45 59 AM
> 
> We'll want to make sure that the hamburger menu is showing for any screen size that is too small to display all of the menu items without overlapping. Screenshot 2024-06-03 at 7 44 15 AM

woah, @theyokohamalife  I made sure it works for all screen widths on my device....but yours looks smashed 🤔 let me see if I can reproduce

Here it is for the smallest desktop/landscape on mine
image

and here's ultrawide

image

File: components/HamburgerMenu.vue (hunk 1 above)
Comment (ermish) [reply components/HamburgerMenu.vue:39]: done ✅

File: components/HamburgerMenu.vue (hunk 2 above)
Comment (ermish) [reply components/HamburgerMenu.vue:71]: makes sense to me. I'll add it later tonight

File: components/HamburgerMenu.vue (hunk 4 above)
Comment (ermish) [reply components/HamburgerMenu.vue:7]: good observation. So in Vue, you can reuse the same ids in components as their name will change when being imported and the parent grouping is added so you don't have to prefix your component. like `header` becomes `hamburger-menu-header` when it's in the `hamburger-menu` component. 

butttttt.....looks like our compiler isn't doing this anymore?? 🤔 I blame nuxt 🙄 .  lemme see if we can turn it back on otherwise I'll add the prefix

**Comment by theyokohamalife** [comment]
> woah, @theyokohamalife I made sure it works for all screen widths on my device....but yours looks smashed 🤔 let me see if I can reproduce

You have to manually drag the screen to a smaller size to see where that happens. The rest of the page switches down to the tablet/mobile view, but the hamburger menu isn't showing.

https://github.com/ourjapanlife/findadoc-web/assets/31802656/ead5d1d2-23d1-4bdc-85e6-f28751214835

**Comment by ermish** [comment]
> > woah, @theyokohamalife I made sure it works for all screen widths on my device....but yours looks smashed 🤔 let me see if I can reproduce
> 
> You have to manually drag the screen to a smaller size to see where that happens. The rest of the page switches down to the tablet/mobile view, but the hamburger menu isn't showing.
> 
>  Screen.Recording.2024-06-03.at.9.08.04.AM.mov

that's so strange...mine doesn't do that when I resize 🤔

**Comment by Tch4lla** [comment]
https://github.com/ourjapanlife/findadoc-web/assets/99618731/5a65c506-49d8-4b71-8e71-55ede196e457

Made it look a bit smoother and made the menu pull from the right

**Comment by ermish** [comment]
> hamburger.mp4 
> Made it look a bit smoother and made the menu pull from the right

this is awesome! I didn't even think to use the `Transition` option!

File: i18n/locales/ja.json [hunk 11]
```diff
@@ -86,8 +76,16 @@
     "topNav": {
         "home": "ホーム",
         "about": "私たちについて",
+        "home": "ホーム",
         "submit": "医師情報の追加"
     },
+    "hamburgerMenu": {
+        "about": "私たちについて",
+        "home": "ホーム",
+        "contactus": "お問い合わせ",
```
Comment (theyokohamalife) [inline i18n/locales/ja.json:85]: The key for this string should be the same as what's referenced in the component and also en.json.
```suggestion
        "contact": "お問い合わせ",
```

File: package.json [hunk 12]
```diff
@@ -13,6 +13,7 @@
         "prod:build": "yarn && yarn generate && nuxi generate",
         "prod:start": "nuxi start",
         "prepare:ci": "nuxi prepare",
+        "test": "yarn test:pinia --run && cypress open && yarn test:e2e",
```
Comment (theyokohamalife) [inline package.json:16]: Let's move this to a separate PR since it's not related to the hamburger menu.

File: components/HamburgerMenu.vue [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            enter-from-class="opacity-0 translate-x-2.5" enter-to-class="opacity-100 translate-x-0"
+            leave-from-class="opacity-100 translate-x-0" leave-to-class="opacity-0 translate-x-2.5">
+            <div data-testid="hamburger-menu" v-show="isMenuOpen"
+                class="z-20 absolute top-0 right-0 h-full w-2/3 pt-6 pb-2 rounded bg-primary-bg border-2 flex flex-col justify-between ">
+                <div data-testid="hamburger-upper-section">
+                    <div data-testid="hamburger-header" class="flex justify-between px-5">
+                        <div class="flex">
+                            <div class="title-text flex flex-col flex-shrink-0" data-testid="logo">
+                                <div class="text-lg text-primary-text group-hover:text-primary-hover font-semibold">
+                                    Find a Doc, Japan
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:15]: Non-blocking: We should use a variable here instead of raw text—ideally a global one.

File: components/HamburgerMenu.vue [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                            <SVGGithubIcon data-testid="hamburger-menu-github-icon" alt="hamburger menu" class="h-10 w-8" />
+                        </NuxtLink>
+                        <!-- Netlify Icons are available here: https://www.netlify.com/press/#badges -->
+                        <NuxtLink :to="'https://www.netlify.com'" target="_blank" data-testid="hamburger-menu-footer-dev-links-netlify">
+                            <img src="https://www.netlify.com/v3/img/components/netlify-light.svg"
+                                alt="Deploys by Netlify" class=" h-10 w-20" />
+                        </NuxtLink>
+                    </div>
+                    <div data-testid="hamburger-menu-footer-copyright" class="text-primary-text-muted text-xs">
+                        © {{ new Date().getUTCFullYear() }} Find a Doc, Japan
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:79]: Let's replace this raw text with a variable

File: components/HamburgerMenu.vue [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+                        <div class="mt-0.5">NPO
+                            <NuxtLink :to="'https://www.npo-hiroba.or.jp/search/zoom.php?pk=121289'" target="_blank" class="underline">#9011005010215</NuxtLink>
```
Comment (theyokohamalife) [inline components/HamburgerMenu.vue:81]: ♻️

File: components/TopNav.vue [hunk 16]
```diff
@@ -1,40 +1,52 @@
...
-                            Japan
-                        </div>
+        class="p-4 pl-5 border border-secondary-bg/20 flex justify-between items-center bg-gradient-to-t from-secondary-bg/30 via-primary-bg to-primary-bg">
+        <div id="mobile-site-icon"
+            class="landscape:hidden font-semibold text-xl group transition-colors flex justify-between items-start">
+            <SVGSiteLogo role="img" title="site icon"
+                class="mt-1 mr-1 w-10 h-10 flex-shrink-0 align-middle fill-primary group-hover:fill-primary-hover" />
+            <div class="title-text flex flex-col flex-shrink-0" data-testid="portrait-logo">
+                <div class="text-lg text-primary group-hover:text-primary-hover">
+                    Find a Doc,
```
Comment (theyokohamalife) [inline components/TopNav.vue:10]: ♻️

File: components/TopNav.vue [hunk 17]
```diff
@@ -1,40 +1,52 @@
...
+                </div>
+                <div class="text-sm text-primary leading-none group-hover:text-primary-hover">
+                    Japan
```
Comment (theyokohamalife) [inline components/TopNav.vue:13]: ♻️

File: components/TopNav.vue [hunk 18]
```diff
@@ -1,40 +1,52 @@
...
+                </div>
+            </div>
+        </div>
+        <div id="desktop-site-icon" class="portrait:hidden font-semibold text-xl group transition-colors items-start">
+            <NuxtLink class="flex" to="/">
+                <SVGSiteLogo role="img" title="site icon"
+                    class="mr-1 w-10 h-10 flex-shrink-0 align-middle fill-primary group-hover:fill-primary-hover" />
+                <div class="title-text flex flex-col flex-shrink-0" data-testid="landscape-logo">
+                    <div class="text-lg text-primary group-hover:text-primary-hover">
+                        Find a Doc
```
Comment (theyokohamalife) [inline components/TopNav.vue:23]: ♻️

File: components/TopNav.vue [hunk 19]
```diff
@@ -1,40 +1,52 @@
...
+                    </div>
+                    <div class="text-sm text-primary leading-none group-hover:text-primary-hover">
+                        Japan
```
Comment (theyokohamalife) [inline components/TopNav.vue:26]: ♻️

File: cypress/e2e/about.cy.ts [hunk 20]
```diff
@@ -7,73 +7,73 @@ describe('About page', () => {
             cy.visit('/about')
         })
 
-        it('shows the desktop nav', () => {
-            cy.get('[data-testid="desktop-nav"]').should('exist')
-        });
+        it('shows the desktop top nav', () => {
+            cy.get('[data-testid="landscape-searchbar"]').should('exist')
```
Comment (theyokohamalife) [inline cypress/e2e/about.cy.ts:11]: Since the element will always be present in the DOM with these changes, the test should confirm that it's not visible.
```suggestion
            cy.get('[data-testid="landscape-searchbar"]').should("not.be.visible")
```

**Comment by theyokohamalife** [comment]
Nice work with the tests! I left a couple small suggestions mainly about replacing the raw text with a variable. It's good practice since the same text is used repeatedly throughout the component.

File: package.json (hunk 12 above)
Comment (ermish) [reply package.json:16]: I would pitch this is a refactoring improvement within the tests area I was touching 😎  
(but I can move it if I must 😿 )

File: components/TopNav.vue (hunk 17 above)
Comment (ermish) [reply components/TopNav.vue:13]: I don't think this one should be as it's part of a visual logo and it is split into parts.
It also doesn't change with translations

File: components/TopNav.vue (hunk 16 above)
Comment (ermish) [reply components/TopNav.vue:10]: I don't think this one should be as it's part of a visual logo and it is split into parts.
It also doesn't change with translations

File: components/TopNav.vue (hunk 18 above)
Comment (ermish) [reply components/TopNav.vue:23]: I don't think this one should be as it's part of a visual logo and it is split into parts.
It also doesn't change with translations

File: components/TopNav.vue (hunk 19 above)
Comment (ermish) [reply components/TopNav.vue:26]: I don't think this one should be as it's part of a visual logo and it is split into parts.
It also doesn't change with translations

File: cypress/e2e/about.cy.ts (hunk 20 above)
Comment (ermish) [reply cypress/e2e/about.cy.ts:11]: good call

**Comment by ermish** [comment]
@theyokohamalife Ready again for review!

**Comment by ermish** [comment]
@SocketSecurity ignore npm/esbuild@0.20.2
@SocketSecurity ignore npm/@nuxt/telemetry@2.5.4

**Comment by ermish** [comment]
@theyokohamalife  FYI: Failing test is unrelated and part of the mod form. I'd like to get this merged and have the test fixed separate

File: cypress/e2e/moderationDashboard.cy.ts [hunk 21]
```diff
@@ -41,9 +41,9 @@ describe(
                     )
             });
 
-            it("it shows the moderation top nav", () => {
+            it.skip("it shows the moderation top nav", () => {
```
Comment (theyokohamalife) [inline cypress/e2e/moderationDashboard.cy.ts:44]: Approving the PR, but I think we should remove the `.skip` before merging?

**Review by theyokohamalife — APPROVED** [review]

File: cypress/e2e/moderationDashboard.cy.ts (hunk 21 above)
Comment (ermish) [reply cypress/e2e/moderationDashboard.cy.ts:44]: since this one is failing unrelated to this PR, let's fix it separately. (I tried initially and there wasn't a quick fix)

Comment (ermish) [reply cypress/e2e/moderationDashboard.cy.ts:44]: https://github.com/ourjapanlife/findadoc-web/issues/540 created!
